fix(all): delay creation of child elements to follow w3c guidelines

peitschie opened the pull request:
To provide a bit more context, I need the fully created web-component (with children) available immediately after creation, hence why I've also exposed the `auInit` method that I am now calling manually.

I don't particularly like this solution, so am really putting it up for comments/discussion (and so people know where to look for the problem!)

bigopon commented on src/custom-element-utilities.ts:
```diff
@@ -50,6 +50,7 @@ export const createWebComponentClassFromBehavior = (
     _bindings: Binding[];
     /**@internal */
     au: { controller: Controller; };
+    auInit: () => void;
```
Let's mark this with @internal as well

bigopon commented on src/custom-element-utilities.ts:
```diff
@@ -77,20 +78,26 @@ export const createWebComponentClassFromBehavior = (
-      const controller = behavior.create(childContainer, behaviorInstruction, this, bindings);
-      controller.created(null);
+      let initialised = false;
+      this.auInit = () => {
```
I don't think we need to define it per instance like this to avoid memory pressure. It can be totally a class method. or a loose function. About intialization check, we can use a weakset for it.

bigopon commented:
> I don't particularly like this solution, so am really putting it up for comments/discussion (and so people know where to look for the problem!)

I don't think there's anything wrong with the approach chosen in this PR, it's pretty much the only way. Custom element v1 has issue with no reliable entry point. Note that we still need to handle it in the attribute changed callback too, not just only connected callback, but we need to verify it via some convoluted tests probably to have the broken behavior tested. I think we can probably just follow some well known lib in this space like hyperhtml.

About this PR, can you help remove dist files? And it would be nice if you could help add a few more tests, since you are already helping a lot.

> To provide a bit more context, I need the fully created web-component (with children) available immediately after creation, hence why I've also exposed the auInit method that I am now calling manually.

This would be a nice doc addition

peitschie commented:
Yep. I'll tackle those fixes & changes next week.

> This would be a nice doc addition

Is it worth leaving this method public in that case, so other users can call it in a similar use-cases?

bigopon commented:
@peitschie from what i see, its an internal bootstrapping this module provides to avoid quirks of web component custom elements so we dont need to export it.

What we can do I think is to give notes about that and why we do it in the doc

peitschie commented:
@bigopon I have added two tests that exercise this new behaviour to ensure the `document.createElement` calls now succeed. I've also added some lines to the documentation to note the modified behaviour... please let me know if you feel more is good to have 😸 

The remaining niggle is that I have an additional test which I did not include because it fails... but I can't explain why. The test is:
```
it('should initialise web component when attribute changed', async () => {
  class CustomElementThree {
    static $view = '<template>${testing}</template>';
    static $resource = {
      bindables: ['testing']
    }
  }
  const { registry, dispose } = await bootstrapAurelia({
    root: class RootViewModel {
      static $view = '<template></template>';
    },
    resources: [CustomElementThree]
  });
  registry.register(CustomElementThree);
  await waitForTimeout(50);

  const customComponent = document.createElement("custom-element-three");
  customComponent.setAttribute("testing", "One");
  expect(customComponent.textContent).toBe('One');
  dispose();
});
```

The output of this is:
```
SUMMARY:
√ 2 tests completed
× 1 test failed

FAILED TESTS:
  ce.integration.spec.ts
    × should initialise web component when attribute changed
      HeadlessChrome 74.0.3729 (Windows 10.0.0)
    Error: Expected ' ' to be 'One'.
        at <Jasmine>
        at UserContext.<anonymous> (webpack:///test/ce.integration.spec.ts:60:40 <- test/ce.integration.spec.js:19171:45)
```

What's interesting is that the value of the custom element has changed (from `''`, i.e. a 0-length string, to `' '`, i.e., a string of a single space), so setting the attribute *is* having an effect on things... but it's not setting the value to `'One'` like I would expect.

I assume I'm missing something with the resource setup here, but I can't find where. Any ideas @bigopon ?

Thanks for the review work and fantastically useful plugin!

peitschie commented:
Ok... figured out the last hiccup. The view bindings have no effect on the document content until the node is connected. So, calling `setAttribute` initialises things, but the new attribute value won't be reflected until the custom element is appended to the host node. Then, it picks up the specified custom attribute value!

😅

bigopon commented:
@peitschie I don't think change handler for attribute change is synchronous, you are setting the attribute and validate the result immediately, which should fail. That's why i had a timeout, because timing is not observable from us and we have to go totally reactive here i guess

bigopon commented on README.md:
```diff
@@ -142,5 +158,6 @@ To run the unit tests, first ensure that you have followed the steps above in or
-* For each view model class, a corresponding native custom element class will be created and defined, with the name derived from metadata and fallbacks to view model class name. If there is hyphen `-` in the name of a custom element view model, a prefix (`au-` by default) will be added to the name. This can be change in `CustomElementRegistry` instance.
+* For each view model class, a corresponding native custom element class will be created and defined, with the name derived from metadata and fallbacks to view model class name. If there is no hyphen `-` in the name of a custom element view model, a prefix (`au-` by default) will be added to the name. This can be change in `CustomElementRegistry` instance.
 * Slot: By default, content projection is done using Aurelia slot emulation. This is to keep it consistent with the rest of Aurelia ecosystem. To use native slot/shadow dom for content projection, decorate view model class with `@useShadowDOM`.
+* To comply with Custom element v1 specs, the element created by by Aurelia when calling `document.createElement` is empty until an attribute is modified or the element is added to a document. Specifically, the child elements are not created until the `connectedCallback` or `attributeChangedCallback` hooks are triggered.
```
Awesome 👍

bigopon commented on src/custom-element-utilities.ts:
```diff
@@ -50,9 +50,16 @@ export const createWebComponentClassFromBehavior = (
     /**@internal */
     au: { controller: Controller; };
+    /**@internal */
+    initialised = false;
```
let's move initialization to constructor body, it's easier when you don't have to read the whole class body to know what's available

peitschie replied:
Done!

peitschie commented:
Hiya @bigopon , just wanted to confirm in case this slipped off your radar, this is ready for another review when you get a chance 😺 (absolutely no rush from my end!)

bigopon commented:
@peitschie i thought you were still going. Cool ill have a look. Have you addressed my comment above with regards to the tests https://github.com/aurelia/web-components/pull/13#issuecomment-494288139

peitschie commented:
@bigopon the issue I hit with the failing test ended up being caused by the attribute handlers not taking effect until after the element is attached.

However, thinking further, I realised I had missed a test case where the attribute value is changed *after* connecting (which I believe is the scenario you were meaning with regards to the delay). I've added this one as well.

I've pushed all updates, so this is good for review!

bigopon commented on test/ce.integration.spec.ts:
```diff
@@ -19,4 +20,91 @@ describe('ce.integration.spec.ts', () => {
+
+    const customComponent = document.createElement("custom-element-three");
+    customComponent.setAttribute("testing", "xx-ignored-xx");
+    expect(customComponent.textContent).toBe('One');
```
Is this test working for you?

bigopon replied:
Nvm, i misread it.

bigopon commented on test/ce.integration.spec.ts:
```diff
@@ -19,4 +20,91 @@ describe('ce.integration.spec.ts', () => {
+    const customComponent = document.createElement("custom-element-four");
+    customComponent.setAttribute("testing", "One");
+    host.appendChild(customComponent);
+    expect(customComponent.textContent).toBe('One');
```
Is this test working for you? It reads like the `attributeChangedCallback` is getting fired immediately?

peitschie replied:
You are correct in your reading, in that `attributeChangedCallback` is called before the `connectedCallback`. That this merely demonstrates that the `attributeChangedCallback` has no effect, as the behaviour does nothing until the controller is bound. At the point the controller is bound, it picks up the correct value for this attribute.

I'm definitely open to ways to clarify this better... it's really just exercising the various combinations of things and ensuring nothing bad happens 😺

bigopon replied:
If so, can we do
```diff
    const customComponent = document.createElement("custom-element-four");
    customComponent.setAttribute("testing", "One");
+ expect(customComponent.textContent).toBe('');
    host.appendChild(customComponent);
    expect(customComponent.textContent).toBe('One');
```

peitschie replied:
Ok... this assert now reveals something slightly odd. The `textContent` becomes `" "` (i.e., a single space).

That is:
1. Before the attribute is set, `textContent` is `""`
2. After the attribute is set, `textContent` is `" "`
3. After the element is attached, `textContent` is `"One"`

I don't have a good explanation for this. I've added the asserts here, and put a `// TODO` comment there to investigate. At least then you can review, and I'll dig into this further tomorrow perhaps...

bigopon replied:
wow that's odd. A `// todo` will be fine, we will tackle it ... later :D

bigopon commented on test/ce.integration.spec.ts:
```diff
@@ -19,4 +20,91 @@ describe('ce.integration.spec.ts', () => {
+    registry.register(CustomElementFive);
+    await waitForTimeout(50);
+
+    const customComponent = document.createElement("custom-element-four");
```
The CE name is `custom-element-five`

peitschie replied:
Good catch! 😳

bigopon commented on README.md:
```diff
@@ -70,11 +70,12 @@ export function configure(aurelia) {
       class MyCarousel {
         // ...
       }
+      aurelia.use.globalResources(MyCarousel);
```
Should it be `registry.register(MyCarousel)`?

bigopon replied:
Oh, I see the below changes. What is the intention of adding both of them? You wanna show it's possible to have both available in an application?

peitschie replied:
In my tests, `registry.register(MyCarousel)` only works if the element has a defined behaviour (i.e., is registered first as a `globalResource`). I assumed this was intentional, so updated the docs accordingly.

Apologies if I have made the wrong choice!

bigopon replied:
Did you return/await the register `Promise`? I think there could be something wrong, since it shouldn't need anything declared global at all. Webcomponent plugin can run independently

bigopon commented on test/ce.integration.spec.ts:
```diff
@@ -19,4 +20,91 @@ describe('ce.integration.spec.ts', () => {
+    host.appendChild(customComponent);
+    await waitForTimeout(50);
+
+    customComponent.setAttribute("testing", "One");
```
lets add an assertion here that text content should not be `'One'`

bigopon replied:
I think doing that, it would help communicate the timing clearer

peitschie replied:
Done.

bigopon commented:
@EisenbergEffect I think this is good to go. I'll add some tweaks/fixes after this gets merged

bigopon commented:
Cool, merged.

Thanks heap @peitschie
